fix(styles): change focused fields color scheme

TDA opened the pull request:
Fixes #3384 
### Before
<img width="406" alt="screenshot 2016-05-11 14 34 47" src="https://cloud.githubusercontent.com/assets/432707/15197355/e021d54e-1785-11e6-9b9b-d45a3e219fbc.png">

### After
* Has Focus (sign-up page)
<img width="407" alt="screenshot 2016-05-11 14 19 18" src="https://cloud.githubusercontent.com/assets/432707/15197040/29961dd6-1784-11e6-8506-330ce0a69a58.png">

* Has Focus, unmasked (sign-up page)
<img width="395" alt="screenshot 2016-05-11 14 19 27" src="https://cloud.githubusercontent.com/assets/432707/15197079/5a86fc44-1784-11e6-8fb3-5d47b55dbc08.png">

* Has Focus (change password page)
<img width="625" alt="screenshot 2016-05-11 14 18 39" src="https://cloud.githubusercontent.com/assets/432707/15197062/4aac20ec-1784-11e6-87cd-f6a8483a7ce6.png">

* Has Focus, unmasked  (change password page)
<img width="506" alt="screenshot 2016-05-11 14 18 01" src="https://cloud.githubusercontent.com/assets/432707/15197239/2a80ecde-1785-11e6-9dad-a10bf36f69e8.png">


* No Focus, masked
<img width="401" alt="screenshot 2016-05-11 14 36 32" src="https://cloud.githubusercontent.com/assets/432707/15197334/c6ab0dc4-1785-11e6-8ff1-3d467d253bc2.png">

* No Focus, unmasked
<img width="395" alt="screenshot 2016-05-11 14 35 25" src="https://cloud.githubusercontent.com/assets/432707/15197319/b023c1c2-1785-11e6-963d-d89ff75d3e63.png">


@ryanfeeley 
@shane-tomlinson 
@johngruen

ryanfeeley commented:
Wow, looking good! Can I get a demo tomorrow?

On Wednesday, 11 May 2016, Sai Prashanth Chandramouli <
notifications@github.com> wrote:

> Fixes #3384 <https://github.com/mozilla/fxa-content-server/issues/3384>
> Before
>
> [image: screenshot 2016-05-11 14 34 47]
> <https://cloud.githubusercontent.com/assets/432707/15197355/e021d54e-1785-11e6-9b9b-d45a3e219fbc.png>
> After
>
>    -
>
>    Has Focus (sign-up page)
>    [image: screenshot 2016-05-11 14 19 18]
>    <https://cloud.githubusercontent.com/assets/432707/15197040/29961dd6-1784-11e6-8506-330ce0a69a58.png>
>    -
>
>    Has Focus, unmasked (sign-up page)
>    [image: screenshot 2016-05-11 14 19 27]
>    <https://cloud.githubusercontent.com/assets/432707/15197079/5a86fc44-1784-11e6-8fb3-5d47b55dbc08.png>
>    -
>
>    Has Focus (change password page)
>    [image: screenshot 2016-05-11 14 18 39]
>    <https://cloud.githubusercontent.com/assets/432707/15197062/4aac20ec-1784-11e6-87cd-f6a8483a7ce6.png>
>    -
>
>    Has Focus, unmasked (change password page)
>    [image: screenshot 2016-05-11 14 18 01]
>    <https://cloud.githubusercontent.com/assets/432707/15197239/2a80ecde-1785-11e6-9dad-a10bf36f69e8.png>
>    -
>
>    No Focus, masked
>    [image: screenshot 2016-05-11 14 36 32]
>    <https://cloud.githubusercontent.com/assets/432707/15197334/c6ab0dc4-1785-11e6-8ff1-3d467d253bc2.png>
>    -
>
>    No Focus, unmasked
>    [image: screenshot 2016-05-11 14 35 25]
>    <https://cloud.githubusercontent.com/assets/432707/15197319/b023c1c2-1785-11e6-963d-d89ff75d3e63.png>
>
> @ryanfeeley <https://github.com/ryanfeeley>
> @shane-tomlinson <https://github.com/shane-tomlinson>
> @johngruen <https://github.com/johngruen>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/mozilla/fxa-content-server/pull/3727
> Commit Summary
>
>    - fix(styles): change focused fields color scheme
>
> File Changes
>
>    - *M* app/styles/_variables.scss
>    <https://github.com/mozilla/fxa-content-server/pull/3727/files#diff-0>
>    (1)
>    - *M* app/styles/modules/_custom-rows.scss
>    <https://github.com/mozilla/fxa-content-server/pull/3727/files#diff-1>
>    (18)
>    - *M* app/styles/modules/_input-row.scss
>    <https://github.com/mozilla/fxa-content-server/pull/3727/files#diff-2>
>    (5)
>
> Patch Links:
>
>    - https://github.com/mozilla/fxa-content-server/pull/3727.patch
>    - https://github.com/mozilla/fxa-content-server/pull/3727.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/fxa-content-server/pull/3727>
>

TDA commented:
@ryanfeeley Sure thing! I'll catch you on IRC

TDA commented:
@ryanfeeley 
Updated with changes from today's demo.
@shane-tomlinson 
Added some javascript for getting focus/blur and change class accordingly. Didn't want to pollute the JS with CSS properties, so created a separate class and added the style there. Also, using CSS Transitions for faster/smoother performance.

shane-tomlinson commented on app/scripts/views/mixins/floating-placeholder-mixin.js:
```diff
@@ -37,10 +39,20 @@ define(function (require, exports, module) {
       }
     },
 
+    focusFloatingPlaceholder: function (event) {
```
I'm not a fan of passing in the events to functions that do things with the event's currentTarget, the code becomes less generic. I generally use two functions, one which receives the event, and another which receives an element. For example:

```js
events: {
  'blur input': 'onInputBlur',
 ...
},

onInputBlur: function (event) {
  this.unfocusFloatingPlaceholder($(event.currentTarget));
},

unfocusFloatingPlaceholder($inputEl) {
  $inputEl.prev('.label-helper').removeClass('focused');
},
...
```

I find separating the concerns like this makes testing easier and it means the functions that work with the elements can be used elsewhere without passing in an event.

shane-tomlinson commented:
@TDA - some unit tests are needed for this!

shane-tomlinson commented on app/styles/modules/_custom-rows.scss:
```diff
@@ -162,6 +162,24 @@ ul.links {
+  color: $input-row-focus-border-color;
+}
+
+.password[type=text]:focus ~ .show-password-label {
```
Should `.password` be any `input` element?

TDA replied:
I'd say stick with .password, seeing as we don't have .show-password-label after other inputs. I think it'd be better to be as specific as possible.

shane-tomlinson commented:
@TDA - The helper text on age field on the signup page does not have the updated color.

![signup-age-field](https://cloud.githubusercontent.com/assets/848085/15291755/65bc8de4-1b77-11e6-8e1a-54a5677c3148.gif)

@TDA - A couple of questions:

1. How come JS is needed for this? Can the color be applied directly to the `.label-helper` element w/o JS? 
1. What is the difference in these two selectors: `label:focus ~ .input-help-focused` and `.label-helper.focused`? The first was already in use, the second is added in this PR. Can the first be used for the second?

TDA commented:
@shane-tomlinson 
I have re-written the JS as you suggested, with one receiving the event and then calling the other that receives only the element. I will write unit tests for this.

For the other two questions:
1. Discussed in fxa-coord meeting
2. Once again, the existing style is for the helper text *following* the focused element, while here we are trying to achieve a style change for the helper text *before* the focused element. So, the two style declarations are mutually independent.

vladikoff commented on app/styles/modules/_custom-rows.scss:
```diff
@@ -162,6 +162,24 @@ ul.links {
+  color: $show-password-label-color;
+  -webkit-transition: background-color 450ms,
+                      color 450ms;
+  transition: background-color 450ms,
```
@TDA autoprefixer might do this for us already where you only need to set `transition` once.

TDA commented:
@shane-tomlinson Unit tests done. Had to rewrite an older test as well. Ready for review.
Big thanks to @vladikoff for spending over an hour on fixing up the tests.

shane-tomlinson commented on app/scripts/views/mixins/floating-placeholder-mixin.js:
```diff
@@ -30,17 +32,26 @@ define(function (require, exports, module) {
-        $inputEl.prev('.label-helper').text(placeholder).animate( {'top': '-17px'}, 400);
+      // The check for existence is because of the strict equality check,
+      // if placeholder is undefined, then it should not go into the block
+      if (placeholder && placeholder !== '') {
```
> if placeholder is undefined, then it should not go into the block

If the placeholder is undefined, then it *will* go into the block, the strict equality check matches against the empty string only.

TDA replied:
if (placeholder) will be false if placeholder is undefined. `undefined` is falsy, and using the `&&` short-circuits this.

shane-tomlinson commented:
The sass-linter is complaining:

```
app/styles/modules/_custom-rows.scss
  192:3  warning  Expected `transition`, found `z-index`  property-sort-order
  193:3  warning  Expected `z-index`, found `transition`  property-sort-order

app/styles/modules/_input-row.scss
  105:5  warning  Expected `color`, found `opacity`  property-sort-order
  106:5  warning  Expected `opacity`, found `color`  property-sort-order
```

shane-tomlinson commented:
@ryanfeeley - Some questions for you:

* Should the placeholder text turn blue when focused as well? Currently the placeholder text remains grey.

<img width="360" alt="screen shot 2016-05-17 at 14 23 19" src="https://cloud.githubusercontent.com/assets/848085/15326108/9e5186a0-1c44-11e6-81f0-1c4a689f8888.png">

* Should "Must be at least 8 characters" be blue on the change password screen? That same text is blue on the signup page (as shown above):

<img width="241" alt="screen shot 2016-05-17 at 15 29 58" src="https://cloud.githubusercontent.com/assets/848085/15326133/b744a4a8-1c44-11e6-943f-5552e3c9c047.png">

* Should "Must be at least 8 characters" be blue on the complete reset password screen? That same text is blue on the signup page (as shown above):

<img width="437" alt="screen shot 2016-05-17 at 15 37 04" src="https://cloud.githubusercontent.com/assets/848085/15326329/77ced496-1c45-11e6-9d6f-c55c4785b0cf.png">

* Do you think there is the possibility of causing user confusion between links and the helper text on the change password screen? Other than size, they look very similar:

<img width="167" alt="screen shot 2016-05-17 at 15 30 27" src="https://cloud.githubusercontent.com/assets/848085/15326166/dddc066a-1c44-11e6-997d-92938c252887.png">

* Should the reset password screen have a helper label (it does not):

before entering email (no helper, but not supposed to have one):
<img width="399" alt="screen shot 2016-05-17 at 15 36 37" src="https://cloud.githubusercontent.com/assets/848085/15326374/a8d8edba-1c45-11e6-8f61-aeb235f03dc1.png">

after entering email (still no helper):
<img width="383" alt="screen shot 2016-05-17 at 15 37 30" src="https://cloud.githubusercontent.com/assets/848085/15326379/b0470672-1c45-11e6-9ccf-cd1c07174dcf.png">

ryanfeeley commented:
>  Should the placeholder text turn blue when focused as well? Currently the placeholder text remains grey.

No, because there will likely be a link to "Need inspiration" in there and underlines are not great in Fira Sans.

> Should "Must be at least 8 characters" be blue on the change password screen? That same text is blue on the signup page (as shown above):

Same as above.

> Should "Must be at least 8 characters" be blue on the complete reset password screen? That same text is blue on the signup page (as shown above):

Same as above.

> Do you think there is the possibility of causing user confusion between links and the helper text on the change password screen? Other than size, they look very similar:

I worry about that but they are only blue when focused and users are familiar with "Forgot password?" text as links.


> Should the reset password screen have a helper label (it does not):

No, it's the only field on the page. Helper labels help disambiguate multiple forms, but are applied on an as-needed basis. The best example of which is the old password, repeat password forms.

shane-tomlinson commented:
Thanks @ryanfeeley for the responses. Assigning to Sai to fix up the lint tests.

shane-tomlinson commented:
Thanks @ryanfeeley for the responses. Assigning to Sai to fix up the lint tests.

pdehaan commented:
I probably lack a lot of context to say anything useful, but scrolling through all the cropped screenshots above, it seems a bit weird that sometimes "Must be at least 8 characters" is sometimes left aligned and sometimes center aligned.

ryanfeeley commented:
> I probably lack a lot of context to say anything useful

Careless talk costs Ryan time.

ryanfeeley commented:
@pdehaan  ![5023201644_63ecf23805_b](https://cloud.githubusercontent.com/assets/68213/15330844/896f7e26-1c2c-11e6-9a8b-20cf5df448d2.jpg)

pdehaan commented:
> Careless talk costs Ryan time.

![](https://i.kinja-img.com/gawker-media/image/upload/xxlds8a6iai47xyul3l3.png)

TDA commented:
@shane-tomlinson Linted :) Ready for final r  :)
